zulu/datetime.py

```diff
@@ -379,6 +380,154 @@ def replace(self,
+        """This method helps in shifting the values to end of the day."""
+        return self.shift(days=count * 1,
+                          hours=-self.hour,
+                          minutes=-self.minute,
+                          seconds=-self.second,
+                          microseconds=-1
+                          )
+
+    def end_of_hour(self, count):
+        """This method helps in shifting the values to end of the hour."""
+        return self.shift(hours=count * 1,
+                          minutes=-self.minute,
+                          seconds=-self.second,
+                          microseconds=-1)
+
+    def end_of_minute(self, count):
+        """This method helps in shifting the values to end of the minute."""
+        return self.shift(minutes=count * 1,
+                          seconds=-self.second,
+                          microseconds=-1)
+
+    def end_of_second(self, count):
+        """This method helps in shifting the values to end of the second."""
+        return self.shift(seconds=count * 1,
+                          microseconds=-1)
+
+    def validate_frame(self, frame):
+        """Validates the given time frame."""
+        attrs = ['year',
+                 'month',
+                 'day',
+                 'hour',
+                 'minute',
+                 'second']
+
+        if frame not in attrs:
+            raise AttributeError('The given time frame {0} is invalid'
+                                 .format(frame))
+
+    def start_of(self, frame):
+        """This method returns a floor value of datetime for the given time
+        frame.
+
+        Args:
+            frame (str): A time frame. Ex: year, month, day, minute, etc.
+
+        Returns:
+            zulu.Datetime.datetime
+        """
+        self.validate_frame(frame)
+
+        method = getattr(self, 'start_of_{0}'.format(frame), None)
+        if callable(method):
+            return method()
+
+    def end_of(self, frame, count=1):
+        """This method returns a ceil value of datetime for the given time
+        frame.
+
+        Args:
+            frame (str): A time frame. Ex: year, month, day, minute, etc.
+            count (int): Number of frames to span
+
+        Returns:
+            zulu.Datetime.datetime
+        """
+        self.validate_frame(frame)
+
+        method = getattr(self, 'end_of_{0}'.format(frame), None)
+        if callable(method):
+            return method(count=count)
+
+    def span(self, frame, count=1):
+        """Returns two new :class:`Datetime <datetime.datetime>` objects,
+        representing the time span of the :class:`Datetime <datetime.datetime>`
+         object in a given time frame.
+
+        Args:
+             frame (str): A time frame. Ex: year, month, day, minute, etc.
+             count (int): Number of frames to span
+
+        Returns:
+            floor: <datetime.datetime>
+            ceil: <datetime.datetime>
+        """
+        self.validate_frame(frame)
```

Comment:
I don't understand. `start_of` and `end_of` are being called here and they in turn immediately call `validate_frame`. That's 3 total calls to `validate_frame`. Removing this one makes it 2 total (still not ideal but at least one less). May need some refactoring later to ensure only one validation (obviously minor optimization) but not a big deal right now.